{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "import pysam \n",
    "import pandas as pd\n",
    "import sys\n",
    "sys.path.append(\"./scripts\")\n",
    "import utils"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "source": [
    "df = pd.read_csv(\"/Users/zluo/TestData/GENCODE/GRCh38/gencode.v31.primary_assembly.protein_coding.exon.srt.bed\",sep='\\t', names=['chrom','start','end','name','score','strand'])"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "source": [
    "fa = pysam.FastaFile(\"/Users/zluo/TestData/GENCODE/GRCh38/Sequence/GRCh38.p13.genome.fa\")"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "gtf_gene, gtf_ts, gtf_exon, gtf_CDS = utils.parse_gtf(\"../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.gtf\")"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Dask Apply: 100%|██████████| 24/24 [00:24<00:00,  1.03s/it]\n",
      "Dask Apply: 100%|██████████| 24/24 [04:15<00:00, 10.65s/it]\n",
      "Dask Apply: 100%|██████████| 24/24 [27:20<00:00, 68.36s/it]\n",
      "Dask Apply: 100%|██████████| 24/24 [16:25<00:00, 41.08s/it]\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "source": [
    "gtf_gene.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.gene.tab',sep='\\t',index=False)\n",
    "gtf_ts.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.ts.tab',sep='\\t',index=False)\n",
    "gtf_exon.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.exon.tab',sep='\\t',index=False)\n",
    "gtf_CDS.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.CDS.tab',sep='\\t',index=False)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "source": [
    "gtf_gene = pd.read_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.gene.tab',sep='\\t')\n",
    "gtf_ts = pd.read_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.ts.tab',sep='\\t', dtype={'transcript_support_level':str})\n",
    "gtf_exon = pd.read_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.exon.tab',sep='\\t')\n",
    "gtf_CDS = pd.read_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.CDS.tab',sep='\\t')"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "source": [
    "# assign cds to transcripts\n",
    "gtf_ts = utils.assign_cds_to_ts(gtf_ts, gtf_CDS)\n",
    "gtf_ts.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.ts.tab',sep='\\t',index=False)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "gtf_gene_coding = gtf_gene[gtf_gene['gene_type'] == 'protein_coding']\n",
    "gtf_ts_coding = gtf_ts[gtf_ts['gene_id'].isin(gtf_gene_coding['gene_id'])]\n",
    "gtf_ts_coding_tsl1and2 = gtf_ts_coding[gtf_ts_coding['transcript_support_level'].isin(['1','2',1,2])]\n",
    "gtf_exon_coding_tsl1and2 = gtf_exon[gtf_exon['transcript_id'].isin(gtf_ts_coding_tsl1and2['transcript_id'])]"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "source": [
    "gtf_gene_coding\n",
    "gtf_ts_coding_tsl1and2_sample = gtf_ts_coding_tsl1and2.sample(n=1000, random_state=1992)\n",
    "with open('/Users/zluo/TestData/GENCODE/GRCh38/gtf_ts_coding_tsl1and2_sample1000_intron.bed', 'w') as intron_bed:\n",
    "    for idx, row in gtf_ts_coding_tsl1and2_sample.iterrows():\n",
    "        ts_id = row['transcript_id']\n",
    "        exons  = gtf_exon_coding_tsl1and2[gtf_exon_coding_tsl1and2['transcript_id'] == ts_id]\n",
    "        exons = exons.sort_values('exon_number')\n",
    "        exon_cnt = exons['exon_number'].max()\n",
    "        if exon_cnt < 2:\n",
    "            continue\n",
    "        for idx_exon, exon in exons.iterrows():\n",
    "            exon_number = exon['exon_number']\n",
    "            if exon_number == 1:\n",
    "                exon_end = int(exon['end'])\n",
    "                continue\n",
    "            else:\n",
    "                exon_start = int(exon['start'])\n",
    "                intron_start = exon_end\n",
    "                intron_end = exon_start\n",
    "                exon_end = int(exon['end'])\n",
    "            intron_chrom = exon['seqname']\n",
    "            intron_name = '{}:{}:{}'.format(exon['gene_name'],exon['transcript_id'],exon_number-1)\n",
    "            intron_score = row['transcript_support_level']\n",
    "            intron_strand = exon['strand']\n",
    "            intron_bed.write('{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(intron_chrom, intron_start, intron_end, intron_name, int(intron_score), intron_strand))"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "source": [
    "def get_alt_exons(gtf_ts, gtf_exon):\n",
    "    gtf_exon = gtf_exon.to_dict('index')\n",
    "    gtf_ts = gtf_ts.to_dict('index')\n",
    "    gene_ts_cnt = dict()\n",
    "    exon_usage = dict()\n",
    "    ts_exon_dict = dict()\n",
    "    for idx in gtf_exon:\n",
    "        transcript_id = gtf_exon[idx]['transcript_id']\n",
    "        if transcript_id not in ts_exon_dict:\n",
    "            ts_exon_dict[transcript_id] = list()\n",
    "        ts_exon_dict[transcript_id].append(idx)\n",
    "    \n",
    "    for idx in gtf_ts:\n",
    "        gene_id = gtf_ts[idx]['gene_id']\n",
    "        if gene_id not in gene_ts_cnt:\n",
    "            gene_ts_cnt[gene_id] = 0\n",
    "        gene_ts_cnt[gene_id] += 1\n",
    "    \n",
    "    for transcript_id in ts_exon_dict:\n",
    "        exon_set = sorted(ts_exon_dict[transcript_id])\n",
    "        exon_num = len(exon_set)\n",
    "        for i, exon_idx in enumerate(exon_set):\n",
    "            if i == 0 or i == exon_num - 1:\n",
    "                continue\n",
    "\n",
    "            gene_id = gtf_exon[exon_idx]['gene_id']\n",
    "            exon_id = gtf_exon[exon_idx]['seqname'] + str(gtf_exon[exon_idx]['start']) + str(gtf_exon[exon_idx]['end']) + gtf_exon[exon_idx]['strand']\n",
    "            if exon_id not in exon_usage:\n",
    "                exon_usage[exon_id] = 0\n",
    "            exon_usage[exon_id] += 1\n",
    "    alt_exon_idx = list()\n",
    "    alt_exon_coord = set()\n",
    "    for idx in gtf_exon:\n",
    "        exon_id = gtf_exon[idx]['seqname'] + str(gtf_exon[idx]['start']) + str(gtf_exon[idx]['end']) + gtf_exon[idx]['strand']\n",
    "        gene_id = gtf_exon[idx]['gene_id']\n",
    "        if exon_id in exon_usage and gene_id in gene_ts_cnt:\n",
    "            if exon_usage[exon_id] != gene_ts_cnt[gene_id]:\n",
    "                if exon_id not in alt_exon_coord:\n",
    "                    alt_exon_idx.append(idx)\n",
    "                    alt_exon_coord.add(exon_id)\n",
    "                \n",
    "    return alt_exon_idx"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "source": [
    "exon_idx_list = get_alt_exons(gtf_ts_coding_tsl1and2, gtf_exon)\n",
    "gtf_exon_dict = gtf_exon.to_dict('index')\n",
    "\n",
    "with open('/Users/zluo/TestData/GENCODE/GRCh38/gencode.v31.primary_assembly.coding.alt.exon.bed', 'w') as alt_exon:\n",
    "    for idx in exon_idx_list:\n",
    "        exon = gtf_exon_dict[idx]\n",
    "        # gtf coordinates are 1-based, remember to covert to 0-based.\n",
    "        alt_exon.write('{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(exon['seqname'], exon['start']-1  , exon['end'], exon['exon_id'], '.', exon['strand']))"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "source": [
    "def get_introns(gtf_ts, gtf_exon):\n",
    "    gtf_ts = gtf_ts.to_dict('index')\n",
    "    gtf_exon = gtf_exon.to_dict('index')\n",
    "    ts_exon_dict = dict()\n",
    "    for idx in gtf_exon:\n",
    "        transcript_id = gtf_exon[idx]['transcript_id']\n",
    "        if transcript_id not in ts_exon_dict:\n",
    "            ts_exon_dict[transcript_id] = list()\n",
    "        ts_exon_dict[transcript_id].append(idx)\n",
    "    \n",
    "    for ts_idx in gtf_ts:\n",
    "        transcript_id = gtf_ts[ts_idx]['transcript_id']\n",
    "        exon_set = sorted(ts_exon_dict[transcript_id])\n",
    "        exon_num = len(exon_set)\n",
    "        if exon_num < 2:\n",
    "            continue\n",
    "        intron_chrom = gtf_ts[ts_idx]['seqname']\n",
    "        intron_strand = gtf_ts[ts_idx]['strand']\n",
    "        intron_start = 0\n",
    "        intron_end = 0\n",
    "        for i, exon_idx in enumerate(exon_set):\n",
    "            if i == 0:\n",
    "                if intron_strand == '+':\n",
    "                    intron_start = gtf_exon[exon_idx]['end']\n",
    "                else:\n",
    "                    intron_end = gtf_exon[exon_idx]['start']\n",
    "            else:\n",
    "                if intron_strand == '+':\n",
    "                    intron_end = gtf_exon[exon_idx]['start']\n",
    "                else:\n",
    "                    intron_start = gtf_exon[exon_idx]['end']\n",
    "                intron_name = transcript_id+'_'+str(i)\n",
    "                yield '{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(intron_chrom, intron_start, intron_end, intron_name, '.', intron_strand)\n",
    "                if intron_strand == '+':\n",
    "                    intron_start = gtf_exon[exon_idx]['end']\n",
    "                else:\n",
    "                    intron_end = gtf_exon[exon_idx]['start']"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "source": [
    "with open('/Users/zluo/TestData/GENCODE/GRCh38/gencode.v31.primary_assembly.coding.intron.bed', 'w') as intron_bed:\n",
    "    for intron in get_introns(gtf_ts_coding_tsl1and2, gtf_exon):\n",
    "        intron_bed.write(intron)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "source": [
    "event_stat = pd.read_csv(\"~/TestData/event_stat.tab\", sep='\\t')\n",
    "gtf_ts['start'] = pd.to_numeric(gtf_ts['start'])\n",
    "gtf_ts['end'] = pd.to_numeric(gtf_ts['end'])\n",
    "event_dis_stat = list()\n",
    "for idx, event in event_stat.iterrows():\n",
    "    coord = event['coord']\n",
    "    chrom = coord.split(':')[0]\n",
    "    strand = coord.split(':')[2]\n",
    "    start = int(coord.split(':')[1].split('-')[0])\n",
    "    end = int(coord.split(':')[1].split('-')[1])\n",
    "    protein = event['protein']\n",
    "    #if 'PABPC1' not in protein.split(','):\n",
    "        #continue\n",
    "    target_ts = gtf_ts[gtf_ts['seqname']==chrom]\n",
    "    target_ts = target_ts[target_ts['strand']==strand]\n",
    "    target_ts = target_ts[(target_ts['start']<= start) & (target_ts['end']>=end)]\n",
    "    target_ts = target_ts.head(1)\n",
    "    if strand == '+':\n",
    "        dis_abs =  int(target_ts['end'] - start)\n",
    "        dis_prop = float((start - target_ts['start']) / (target_ts['end'] -target_ts['start']))\n",
    "    else:\n",
    "        dis_abs = int(end - target_ts['start'])\n",
    "        dis_prop = float((end - target_ts['start']) / (target_ts['end'] -target_ts['start']))\n",
    "    event_dis_stat.append([dis_abs, dis_prop])\n",
    "event_dis_stat = pd.DataFrame(event_dis_stat, columns=['dis_abs','dis_prop'])"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "source": [
    "import seaborn as sns\n",
    "\n",
    "ax = sns.displot(x='dis_prop',  data=event_dis_stat)\n",
    "#_ = ax.set_xticklabels(ax.get_xticklabels(),rotation=90)\n"
   ],
   "outputs": [
    {
     "output_type": "display_data",
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWAAAAFgCAYAAACFYaNMAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAS40lEQVR4nO3dfaxld13v8fenrTwoxbZ0nDTD9A4P9aEi1jpiRe69xRpSGnSKQilRGU1lUEGvlxsTtMm1uYaEm6j3BoPAKE1bwkMLgh2lFqFU6gMFhwdLQZGKrZ22dKbgLWoVafneP/YaPJmeztkzs9f+7nP2+5XsnLV+a+21v7+c00/X/PZav5WqQpI0f8d1FyBJy8oAlqQmBrAkNTGAJamJASxJTU7oLuBYnH/++XX99dd3lyFJa8lqjev6DPi+++7rLkGSjtq6DmBJWs8MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWtKFt2Xo6SWby2rL19JnWtq7nA5aktdy9705e9Ma/mMmxrn7ZM2dynIM8A5akJgawJDUxgCWpiQEsSU0MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlLZRZPsMtSXd3DstnwklaKLN8hhvM/jlus+QZsCQ1MYAlqYkBLElNRgvgJFuT3Jjk00k+leS/De2nJHlfks8OP08e2pPktUluS3JLkrPHqk2SFsGYZ8APAv+jqs4EzgFenuRM4FXADVV1BnDDsA7wXOCM4bULeP2ItUlSu9ECuKruqaqPDcv/BPw1sAXYAVw57HYlcOGwvAO4qiZuBk5KctpY9UlSt7mMASfZBnwX8GFgc1XdM2z6PLB5WN4C3LnibfuGtkOPtSvJ3iR7Dxw4MF7RkjSy0QM4yeOA3wN+saq+tHJbVRVQR3K8qtpdVduravumTZtmWKkkzdeoAZzk65iE71uq6l1D870HhxaGn/uH9ruArSve/sShTZI2pDGvggjwJuCvq+o3V2zaA+wclncC165of8lwNcQ5wP0rhiokacMZ81bk7wd+Avhkkk8Mbb8CvAa4JsklwB3ARcO264ALgNuAB4CfGrE2SWo3WgBX1Z8BjzQTxnmr7F/Ay8eqR5IWjXfCSVITA1iSmhjAktTEAJakJgawJDUxgCWpiQEsSU0MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWltSWraeTZCavLVtP7+7OujTmU5ElLbC7993Ji974FzM51tUve+ZMjrNsDGBpRFu2ns7d++6cybGO/7pH89BXvjyTY2kxGMDSiGZ9ljmrYx08nno5BixJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJgawJDUxgCWpiQGsqczy+WE+Q0ya8IkYmsosn+wAPo1BAs+AJamNASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJs4HLOnYHXcCSbqrWHdGC+AklwPPA/ZX1dOGtsuAlwIHht1+paquG7b9MnAJ8BDwC1X13rFqkzRjX31wZhP2L9Nk/WMOQVwBnL9K+/+pqrOG18HwPRO4GPj24T2/neT4EWtbCrN8jJCk2RvtDLiqbkqybcrddwBvr6ovA3+f5DbgGcCHxqpvGczyMULLdFYizUvHl3CvSHJLksuTnDy0bQHuXLHPvqHtYZLsSrI3yd4DBw6stoskrQvzDuDXA08BzgLuAX7jSA9QVburantVbd+0adOMy5Ok+ZlrAFfVvVX1UFV9FfgdJsMMAHcBW1fs+sShTZI2rLkGcJLTVqw+H7h1WN4DXJzk0UmeBJwBfGSetUnSvI15GdrbgHOBU5PsA34VODfJWUABtwMvA6iqTyW5Bvg08CDw8qp6aKzapEeyZevp3L3vzrV3lGZgzKsgXrxK85sOs/+rgVePVY80jVleOQJePaLD81ZkSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNljKAZzlP7patp3d3Z30anqDg70DLbCkfSeQ8uQvAJyhIy3kGLEmLwACWpCYGsCQ1MYAlqYkBfKxm+G2+3+hLy2Upr4KYqRl+mw9+oy8tE8+AJamJASxJTRyCWDTDmLKkjc8AXjTeISYtDYcgJKmJASxJTQxgSWpiAEtSEwNYkpoYwJLUZKoATvL907RJkqY37Rnwb03ZJs3fDCdEkubpsDdiJPk+4JnApiSvXLHp8cDxYxYmTc2bV7ROrXUn3KOAxw37nbii/UvAC8YqSpKWwWEDuKo+CHwwyRVVdcecapKkpTDtXBCPTrIb2LbyPVX1A2MUJUnLYNoAfgfwBuB3gYfGK0eSlse0AfxgVb1+1EokaclMexnaHyT5uSSnJTnl4GvUyiRpg5v2DHjn8POXVrQV8OTZliNJy2OqAK6qJ41diCQtm6kCOMlLVmuvqqtmW44kLY9phyC+Z8XyY4DzgI8BBrAkHaVphyB+fuV6kpOAt49RkCQti6OdjvJfAMeFJekYTDsG/AdMrnqAySQ83wZcM1ZRkrQMph0D/vUVyw8Cd1TVvhHqkaSlMdUQxDApz98wmRHtZODfxyxKkpbBtE/EuAj4CPBC4CLgw0mcjlKSjsG0QxCXAt9TVfsBkmwC3g+8c6zCJGmjm/YqiOMOhu/gC0fwXknSKqY9A74+yXuBtw3rLwKuG6ckSVoOaz0T7qnA5qr6pSQ/Ajxr2PQh4C1jFydJG9laZ8D/F/hlgKp6F/AugCTfMWz7oRFrk6QNba1x3M1V9clDG4e2baNUJElLYq0APukw2x47wzokaemsFcB7k7z00MYkPw18dJySJGk5rDUG/IvAu5P8GP8RuNuBRwHPH7EuSdrwDhvAVXUv8MwkzwaeNjS/p6o+MHplkrTBTTsf8I3AjSPXIklLxbvZJKnJaAGc5PIk+5PcuqLtlCTvS/LZ4efJQ3uSvDbJbUluSXL2WHVJ0qIY8wz4CuD8Q9peBdxQVWcANwzrAM8Fzhheu4DXj1iXJC2E0QK4qm4CvnhI8w7gymH5SuDCFe1X1cTNwElJThurNklaBPMeA95cVfcMy58HNg/LW4A7V+y3b2h7mCS7kuxNsvfAgQPjVSpJI2v7Eq6qiv94ztyRvG93VW2vqu2bNm0aoTJJmo95B/C9B4cWhp8H5xi+C9i6Yr8nDm2StGHNO4D3ADuH5Z3AtSvaXzJcDXEOcP+KoQpJ2pCmnZD9iCV5G3AucGqSfcCvAq8BrklyCXAHk+fLwWRy9wuA24AHgJ8aqy5JWhSjBXBVvfgRNp23yr4FvHysWiRpEXknnCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJgawJDUxgCWpiQEsSU0MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJgawJDUxgCWpiQEsSU0MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJgawJDU5oeNDk9wO/BPwEPBgVW1PcgpwNbANuB24qKr+saM+SZqHzjPgZ1fVWVW1fVh/FXBDVZ0B3DCsS9KGtUhDEDuAK4flK4EL+0qRpPF1BXABf5zko0l2DW2bq+qeYfnzwObV3phkV5K9SfYeOHBgHrVK0ihaxoCBZ1XVXUm+CXhfkr9ZubGqKkmt9saq2g3sBti+ffuq+0jSetByBlxVdw0/9wPvBp4B3JvkNIDh5/6O2iRpXuYewEm+IcmJB5eB5wC3AnuAncNuO4Fr512bJM1TxxDEZuDdSQ5+/lur6vokfwlck+QS4A7goobaJGlu5h7AVfU54DtXaf8CcN6865GkLot0GZokLRUDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJgawJDUxgCWpiQEsSU0MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJgawJDUxgCWpiQEsSU0MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSk4UL4CTnJ/lMktuSvKq7Hkkay0IFcJLjgdcBzwXOBF6c5MzeqiRpHAsVwMAzgNuq6nNV9e/A24EdzTVJ0ihSVd01fE2SFwDnV9VPD+s/AXxvVb1ixT67gF3D6rcAnzmCjzgVuG9G5S4a+7b+bNR+gX071H1Vdf6hjSfMpp75qardwO6jeW+SvVW1fcYlLQT7tv5s1H6BfZvWog1B3AVsXbH+xKFNkjacRQvgvwTOSPKkJI8CLgb2NNckSaNYqCGIqnowySuA9wLHA5dX1adm+BFHNXSxTti39Wej9gvs21QW6ks4SVomizYEIUlLwwCWpCYbMoDXup05yaOTXD1s/3CSbQ1lHpUp+vbKJJ9OckuSG5L8p446j9S0t6An+dEklWTdXOI0Td+SXDT83j6V5K3zrvFoTfH3eHqSG5N8fPibvKCjziOV5PIk+5Pc+gjbk+S1Q79vSXL2UX1QVW2oF5Mv7/4OeDLwKOCvgDMP2efngDcMyxcDV3fXPcO+PRv4+mH5Z9dD36bp17DficBNwM3A9u66Z/g7OwP4OHDysP5N3XXPsG+7gZ8dls8Ebu+ue8q+/RfgbODWR9h+AfBHQIBzgA8fzedsxDPgaW5n3gFcOSy/EzgvSeZY49Fas29VdWNVPTCs3szkWupFN+0t6L8G/G/g3+ZZ3DGapm8vBV5XVf8IUFX751zj0ZqmbwU8flj+RuDuOdZ31KrqJuCLh9llB3BVTdwMnJTktCP9nI0YwFuAO1es7xvaVt2nqh4E7geeMJfqjs00fVvpEib/l150a/Zr+Cfe1qp6zzwLm4FpfmffDHxzkj9PcnOSh92yuqCm6dtlwI8n2QdcB/z8fEob3ZH+t7iqhboOWLOT5MeB7cB/7a7lWCU5DvhN4CebSxnLCUyGIc5l8i+Wm5J8R1X9v86iZuTFwBVV9RtJvg94c5KnVdVXuwtbBBvxDHia25m/tk+SE5j80+gLc6nu2Ex1q3aSHwQuBX64qr48p9qOxVr9OhF4GvAnSW5nMua2Z518ETfN72wfsKeqvlJVfw/8LZNAXnTT9O0S4BqAqvoQ8Bgmk9msdzOZNmEjBvA0tzPvAXYOyy8APlDDyPqCW7NvSb4LeCOT8F0vY4mH7VdV3V9Vp1bVtqraxmRs+4eram9PuUdkmr/H32dy9kuSU5kMSXxujjUerWn69g/AeQBJvo1JAB+Ya5Xj2AO8ZLga4hzg/qq654iP0v1t40jfYF7A5Czi74BLh7b/xeQ/Wpj8EbwDuA34CPDk7ppn2Lf3A/cCnxhee7prnkW/Dtn3T1gnV0FM+TsLkyGWTwOfBC7urnmGfTsT+HMmV0h8AnhOd81T9uttwD3AV5j8C+US4GeAn1nxO3vd0O9PHu3fo7ciS1KTjTgEIUnrggEsSU0MYElqYgBLUhMDWJKaGMCS1MRbkbWhJLkM+GcmE8DcVFXv761IemQGsDakqvqfYxw3yQk1mcBJOmYOQWjdS3Jpkr9N8mfAtwxtVyR5wbD8mhWT1P/6YY5zRZI3JNk7HO95Q/tPJtmT5APADUlOSfL7w/FuTvL0Yb/Lkrw5yYeSfDbJS8fvvdYzz4C1riX5biZzEJzF5O/5Y8BHV2x/AvB84FurqpKctMYhtzGZ5/YpwI1Jnjq0nw08vaq+mOS3gI9X1YVJfgC4avh8gKczmSzoG4CPJ3lPVa2LOXA1f54Ba737z8C7q+qBqvoSD58M5n4mE7i/KcmPAA8ceoBDXFNVX62qzzKZEOdbh/b3VdXBCbqfBbwZoKo+ADwhycFJx6+tqn+tqvuAG5mEubQqA1gb2jBe+wwmTz55HnD9Wm95hPV/mfYj11iXvsYA1np3E3BhkscmORH4oZUbkzwO+Maqug7478B3rnG8FyY5LslTmDzr7DOr7POnwI8Nxz8XuG84+wbYkeQxw9DHuUymbJRW5Riw1rWq+liSq5lMd7ifhwfeicC1SR7DZArBV65xyH9gMkXp45lMPfhvqzwu8DLg8iS3MBnS2Lli2y1Mhh5OBX7N8V8djtNRSoMkVwB/WFXvPMr3Xwb8c1U94pUW0koOQUhSE8+AtXSSXAq88JDmd1TVqzvq0fIygCWpiUMQktTEAJakJgawJDUxgCWpyf8H5elihEgv05sAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 360x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     }
    }
   ],
   "metadata": {}
  }
 ],
 "metadata": {
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.6.13 64-bit ('defcomp': conda)"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  },
  "interpreter": {
   "hash": "ab868fc8bf654ec94d4feddc3281e92c312f95aedf02d0519971ccf21f966c8d"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}