{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam \n",
    "import pandas as pd "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"/Users/zluo/TestData/GENCODE/GRCh38/gencode.v31.primary_assembly.protein_coding.exon.srt.bed\",sep='\\t', names=['chrom','start','end','name','score','strand'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "fa = pysam.FastaFile(\"/Users/zluo/TestData/GENCODE/GRCh38/Sequence/GRCh38.p13.genome.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f3f03d45276d404d8db1d5dac062b188",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dask Apply:   0%|          | 0/24 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6354a0f4a07848dabafc66246807c367",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dask Apply:   0%|          | 0/24 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a69250d844ae43bea4bf7dd4d7ed1fcd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dask Apply:   0%|          | 0/24 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a1bf0db8745c4239975f0c586b663bf4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dask Apply:   0%|          | 0/24 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas\n",
    "import swifter\n",
    "import re\n",
    "def parse_gtf_attrs(feature):\n",
    "    feature_type = feature['feature']\n",
    "    attrs = feature['attribute']\n",
    "    regex = r\"(\\S+\\s\\S+;)\"\n",
    "    attrs_dict = dict()\n",
    "    for attr in re.finditer(regex,attrs):\n",
    "        attr_key = attr[1].split()[0]\n",
    "        attr_val = attr[1].split()[1].rstrip(';').strip('\"')\n",
    "        attrs_dict[attr_key] = attr_val\n",
    "\n",
    "    # parse attributes from gene features\n",
    "    # gene_id, gene_type, gene_name\n",
    "    if feature_type == 'gene':\n",
    "        feature['gene_id'] = attrs_dict['gene_id']\n",
    "        feature['gene_type'] = attrs_dict['gene_type']\n",
    "        feature['gene_name'] = attrs_dict['gene_name']\n",
    "    # parse attributes from transcript features\n",
    "    # gene_id, gene_type, gene_name\n",
    "    if feature_type == 'transcript':\n",
    "        feature['gene_id'] = attrs_dict['gene_id']\n",
    "        feature['gene_type'] = attrs_dict['gene_type']\n",
    "        feature['gene_name'] = attrs_dict['gene_name']\n",
    "        feature['transcript_id'] = attrs_dict['transcript_id']\n",
    "        if 'transcript_support_level' in attrs_dict:\n",
    "            feature['transcript_support_level'] = attrs_dict['transcript_support_level']\n",
    "        else:\n",
    "            feature['transcript_support_level'] = 'NA'\n",
    "    if feature_type == 'exon':\n",
    "        feature['gene_id'] = attrs_dict['gene_id']\n",
    "        feature['gene_type'] = attrs_dict['gene_type']\n",
    "        feature['gene_name'] = attrs_dict['gene_name']\n",
    "        feature['transcript_id'] = attrs_dict['transcript_id']\n",
    "        feature['exon_id'] = attrs_dict['exon_id']\n",
    "        feature['exon_number'] = int(attrs_dict['exon_number'])\n",
    "    if feature_type == 'CDS':\n",
    "        feature['gene_id'] = attrs_dict['gene_id']\n",
    "        feature['gene_type'] = attrs_dict['gene_type']\n",
    "        feature['gene_name'] = attrs_dict['gene_name']\n",
    "        feature['transcript_id'] = attrs_dict['transcript_id']\n",
    "        feature['exon_id'] = attrs_dict['exon_id']\n",
    "        feature['exon_number'] = int(attrs_dict['exon_number'])\n",
    "\n",
    "    return feature\n",
    "\n",
    "\n",
    "def parse_gtf(gtf_file):\n",
    "    gtf = pandas.read_csv(gtf_file, sep='\\t', comment='#', names=['seqname','source','feature','start','end','score','strand','frame','attribute'])\n",
    "    gtf_gene = gtf[gtf['feature']=='gene'].swifter.apply(parse_gtf_attrs,axis=1)\n",
    "    gtf_transcript = gtf[gtf['feature']=='transcript'].swifter.apply(parse_gtf_attrs,axis=1)\n",
    "    gtf_exon = gtf[gtf['feature']=='exon'].swifter.apply(parse_gtf_attrs,axis=1)\n",
    "    gtf_CDS = gtf[gtf['feature']=='CDS'].swifter.apply(parse_gtf_attrs,axis=1)\n",
    "    return gtf_gene, gtf_transcript, gtf_exon,gtf_CDS\n",
    "gtf_gene, gtf_ts, gtf_exon, gtf_CDS = parse_gtf(\"../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gtf_gene' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-e44998f3037f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mgtf_gene\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.gene.tab'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mgtf_ts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.ts.tab'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mgtf_exon\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.exon.tab'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mgtf_CDS\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.CDS.tab'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'gtf_gene' is not defined"
     ]
    }
   ],
   "source": [
    "gtf_gene.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.gene.tab',sep='\\t',index=False)\n",
    "gtf_ts.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.ts.tab',sep='\\t',index=False)\n",
    "gtf_exon.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.exon.tab',sep='\\t',index=False)\n",
    "gtf_CDS.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.CDS.tab',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_gene = pd.read_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.gene.tab',sep='\\t')\n",
    "gtf_ts = pd.read_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.ts.tab',sep='\\t')\n",
    "gtf_exon = pd.read_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.exon.tab',sep='\\t')\n",
    "gtf_CDS = pd.read_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.CDS.tab',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_gene_coding = gtf_gene[gtf_gene['gene_type'] == 'protein_coding']\n",
    "gtf_ts_coding = gtf_ts[gtf_ts['gene_id'].isin(gtf_gene_coding['gene_id'])]\n",
    "gtf_ts_coding_tsl1and2 = gtf_ts_coding[gtf_ts_coding['transcript_support_level'].isin(['1','2',1,2])]\n",
    "gtf_exon_coding_tsl1and2 = gtf_exon[gtf_exon['transcript_id'].isin(gtf_ts_coding_tsl1and2['transcript_id'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_gene_coding\n",
    "gtf_ts_coding_tsl1and2_sample = gtf_ts_coding_tsl1and2.sample(n=1000, random_state=1992)\n",
    "with open('/Users/zluo/TestData/GENCODE/GRCh38/gtf_ts_coding_tsl1and2_sample1000_intron.bed', 'w') as intron_bed:\n",
    "    for idx, row in gtf_ts_coding_tsl1and2_sample.iterrows():\n",
    "        ts_id = row['transcript_id']\n",
    "        exons  = gtf_exon_coding_tsl1and2[gtf_exon_coding_tsl1and2['transcript_id'] == ts_id]\n",
    "        exons = exons.sort_values('exon_number')\n",
    "        exon_cnt = exons['exon_number'].max()\n",
    "        if exon_cnt < 2:\n",
    "            continue\n",
    "        for idx_exon, exon in exons.iterrows():\n",
    "            exon_number = exon['exon_number']\n",
    "            if exon_number == 1:\n",
    "                exon_end = int(exon['end'])\n",
    "                continue\n",
    "            else:\n",
    "                exon_start = int(exon['start'])\n",
    "                intron_start = exon_end\n",
    "                intron_end = exon_start\n",
    "                exon_end = int(exon['end'])\n",
    "            intron_chrom = exon['seqname']\n",
    "            intron_name = '{}:{}:{}'.format(exon['gene_name'],exon['transcript_id'],exon_number-1)\n",
    "            intron_score = row['transcript_support_level']\n",
    "            intron_strand = exon['strand']\n",
    "            intron_bed.write('{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(intron_chrom, intron_start, intron_end, intron_name, int(intron_score), intron_strand))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278267.1\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243485.5\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226944</th>\n",
       "      <td>KI270734.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>131494</td>\n",
       "      <td>137392</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278817.1\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000278817.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC007325.4</td>\n",
       "      <td>ENST00000613204.1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226945</th>\n",
       "      <td>KI270734.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>138082</td>\n",
       "      <td>161750</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277196.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000277196.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC007325.2</td>\n",
       "      <td>ENST00000615165.1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226946</th>\n",
       "      <td>KI270734.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>138082</td>\n",
       "      <td>161852</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277196.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000277196.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC007325.2</td>\n",
       "      <td>ENST00000621424.4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226947</th>\n",
       "      <td>KI270744.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>51009</td>\n",
       "      <td>51114</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278625.1\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000278625.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>RF00026</td>\n",
       "      <td>ENST00000616830.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226948</th>\n",
       "      <td>KI270750.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>148668</td>\n",
       "      <td>148843</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277374.1\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000277374.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>RF00003</td>\n",
       "      <td>ENST00000612925.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>226949 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           seqname   source     feature   start     end score strand frame  \\\n",
       "0             chr1   HAVANA  transcript   11869   14409     .      +     .   \n",
       "1             chr1   HAVANA  transcript   12010   13670     .      +     .   \n",
       "2             chr1   HAVANA  transcript   14404   29570     .      -     .   \n",
       "3             chr1  ENSEMBL  transcript   17369   17436     .      -     .   \n",
       "4             chr1   HAVANA  transcript   29554   31097     .      +     .   \n",
       "...            ...      ...         ...     ...     ...   ...    ...   ...   \n",
       "226944  KI270734.1  ENSEMBL  transcript  131494  137392     .      +     .   \n",
       "226945  KI270734.1  ENSEMBL  transcript  138082  161750     .      -     .   \n",
       "226946  KI270734.1  ENSEMBL  transcript  138082  161852     .      -     .   \n",
       "226947  KI270744.1  ENSEMBL  transcript   51009   51114     .      -     .   \n",
       "226948  KI270750.1  ENSEMBL  transcript  148668  148843     .      +     .   \n",
       "\n",
       "                                                attribute            gene_id  \\\n",
       "0       gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  ENSG00000223972.5   \n",
       "1       gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  ENSG00000223972.5   \n",
       "2       gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  ENSG00000227232.5   \n",
       "3       gene_id \"ENSG00000278267.1\"; transcript_id \"EN...  ENSG00000278267.1   \n",
       "4       gene_id \"ENSG00000243485.5\"; transcript_id \"EN...  ENSG00000243485.5   \n",
       "...                                                   ...                ...   \n",
       "226944  gene_id \"ENSG00000278817.1\"; transcript_id \"EN...  ENSG00000278817.1   \n",
       "226945  gene_id \"ENSG00000277196.4\"; transcript_id \"EN...  ENSG00000277196.4   \n",
       "226946  gene_id \"ENSG00000277196.4\"; transcript_id \"EN...  ENSG00000277196.4   \n",
       "226947  gene_id \"ENSG00000278625.1\"; transcript_id \"EN...  ENSG00000278625.1   \n",
       "226948  gene_id \"ENSG00000277374.1\"; transcript_id \"EN...  ENSG00000277374.1   \n",
       "\n",
       "                                 gene_type    gene_name      transcript_id  \\\n",
       "0       transcribed_unprocessed_pseudogene      DDX11L1  ENST00000456328.2   \n",
       "1       transcribed_unprocessed_pseudogene      DDX11L1  ENST00000450305.2   \n",
       "2                   unprocessed_pseudogene       WASH7P  ENST00000488147.1   \n",
       "3                                    miRNA    MIR6859-1  ENST00000619216.1   \n",
       "4                                   lncRNA  MIR1302-2HG  ENST00000473358.1   \n",
       "...                                    ...          ...                ...   \n",
       "226944                      protein_coding   AC007325.4  ENST00000613204.1   \n",
       "226945                      protein_coding   AC007325.2  ENST00000615165.1   \n",
       "226946                      protein_coding   AC007325.2  ENST00000621424.4   \n",
       "226947                               snRNA      RF00026  ENST00000616830.1   \n",
       "226948                               snRNA      RF00003  ENST00000612925.1   \n",
       "\n",
       "        transcript_support_level  \n",
       "0                            1.0  \n",
       "1                            NaN  \n",
       "2                            NaN  \n",
       "3                            NaN  \n",
       "4                            5.0  \n",
       "...                          ...  \n",
       "226944                       1.0  \n",
       "226945                       1.0  \n",
       "226946                       1.0  \n",
       "226947                       NaN  \n",
       "226948                       NaN  \n",
       "\n",
       "[226949 rows x 14 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_ts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "event_stat = pd.read_csv(\"~/TestData/event_stat.tab\", sep='\\t')\n",
    "gtf_ts['start'] = pd.to_numeric(gtf_ts['start'])\n",
    "gtf_ts['end'] = pd.to_numeric(gtf_ts['end'])\n",
    "event_dis_stat = list()\n",
    "for idx, event in event_stat.iterrows():\n",
    "    coord = event['coord']\n",
    "    chrom = coord.split(':')[0]\n",
    "    strand = coord.split(':')[2]\n",
    "    start = int(coord.split(':')[1].split('-')[0])\n",
    "    end = int(coord.split(':')[1].split('-')[1])\n",
    "    protein = event['protein']\n",
    "    target_ts = gtf_ts[gtf_ts['seqname']==chrom]\n",
    "    target_ts = target_ts[target_ts['strand']==strand]\n",
    "    target_ts = target_ts[(target_ts['start']<= start) & (target_ts['end']>=end)]\n",
    "    target_ts = target_ts.head(1)\n",
    "    if strand == '+':\n",
    "        dis_abs =  int(target_ts['end'] - start)\n",
    "        dis_prop = float((start - target_ts['start']) / (target_ts['end'] -target_ts['start']))\n",
    "    else:\n",
    "        dis_abs = int(end - target_ts['start'])\n",
    "        dis_prop = float((end - target_ts['start']) / (target_ts['end'] -target_ts['start']))\n",
    "    event_dis_stat.append([dis_abs, dis_prop])\n",
    "event_dis_stat = pd.DataFrame(event_dis_stat, columns=['dis_abs','dis_prop'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWAAAAFgCAYAAACFYaNMAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAS40lEQVR4nO3dfaxld13v8fenrTwoxbZ0nDTD9A4P9aEi1jpiRe69xRpSGnSKQilRGU1lUEGvlxsTtMm1uYaEm6j3BoPAKE1bwkMLgh2lFqFU6gMFhwdLQZGKrZ22dKbgLWoVafneP/YaPJmeztkzs9f+7nP2+5XsnLV+a+21v7+c00/X/PZav5WqQpI0f8d1FyBJy8oAlqQmBrAkNTGAJamJASxJTU7oLuBYnH/++XX99dd3lyFJa8lqjev6DPi+++7rLkGSjtq6DmBJWs8MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWtKFt2Xo6SWby2rL19JnWtq7nA5aktdy9705e9Ma/mMmxrn7ZM2dynIM8A5akJgawJDUxgCWpiQEsSU0MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlLZRZPsMtSXd3DstnwklaKLN8hhvM/jlus+QZsCQ1MYAlqYkBLElNRgvgJFuT3Jjk00k+leS/De2nJHlfks8OP08e2pPktUluS3JLkrPHqk2SFsGYZ8APAv+jqs4EzgFenuRM4FXADVV1BnDDsA7wXOCM4bULeP2ItUlSu9ECuKruqaqPDcv/BPw1sAXYAVw57HYlcOGwvAO4qiZuBk5KctpY9UlSt7mMASfZBnwX8GFgc1XdM2z6PLB5WN4C3LnibfuGtkOPtSvJ3iR7Dxw4MF7RkjSy0QM4yeOA3wN+saq+tHJbVRVQR3K8qtpdVduravumTZtmWKkkzdeoAZzk65iE71uq6l1D870HhxaGn/uH9ruArSve/sShTZI2pDGvggjwJuCvq+o3V2zaA+wclncC165of8lwNcQ5wP0rhiokacMZ81bk7wd+Avhkkk8Mbb8CvAa4JsklwB3ARcO264ALgNuAB4CfGrE2SWo3WgBX1Z8BjzQTxnmr7F/Ay8eqR5IWjXfCSVITA1iSmhjAktTEAJakJgawJDUxgCWpiQEsSU0MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWltSWraeTZCavLVtP7+7OujTmU5ElLbC7993Ji974FzM51tUve+ZMjrNsDGBpRFu2ns7d++6cybGO/7pH89BXvjyTY2kxGMDSiGZ9ljmrYx08nno5BixJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJgawJDUxgCWpiQGsqczy+WE+Q0ya8IkYmsosn+wAPo1BAs+AJamNASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJs4HLOnYHXcCSbqrWHdGC+AklwPPA/ZX1dOGtsuAlwIHht1+paquG7b9MnAJ8BDwC1X13rFqkzRjX31wZhP2L9Nk/WMOQVwBnL9K+/+pqrOG18HwPRO4GPj24T2/neT4EWtbCrN8jJCk2RvtDLiqbkqybcrddwBvr6ovA3+f5DbgGcCHxqpvGczyMULLdFYizUvHl3CvSHJLksuTnDy0bQHuXLHPvqHtYZLsSrI3yd4DBw6stoskrQvzDuDXA08BzgLuAX7jSA9QVburantVbd+0adOMy5Ok+ZlrAFfVvVX1UFV9FfgdJsMMAHcBW1fs+sShTZI2rLkGcJLTVqw+H7h1WN4DXJzk0UmeBJwBfGSetUnSvI15GdrbgHOBU5PsA34VODfJWUABtwMvA6iqTyW5Bvg08CDw8qp6aKzapEeyZevp3L3vzrV3lGZgzKsgXrxK85sOs/+rgVePVY80jVleOQJePaLD81ZkSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNljKAZzlP7patp3d3Z30anqDg70DLbCkfSeQ8uQvAJyhIy3kGLEmLwACWpCYGsCQ1MYAlqYkBfKxm+G2+3+hLy2Upr4KYqRl+mw9+oy8tE8+AJamJASxJTRyCWDTDmLKkjc8AXjTeISYtDYcgJKmJASxJTQxgSWpiAEtSEwNYkpoYwJLUZKoATvL907RJkqY37Rnwb03ZJs3fDCdEkubpsDdiJPk+4JnApiSvXLHp8cDxYxYmTc2bV7ROrXUn3KOAxw37nbii/UvAC8YqSpKWwWEDuKo+CHwwyRVVdcecapKkpTDtXBCPTrIb2LbyPVX1A2MUJUnLYNoAfgfwBuB3gYfGK0eSlse0AfxgVb1+1EokaclMexnaHyT5uSSnJTnl4GvUyiRpg5v2DHjn8POXVrQV8OTZliNJy2OqAK6qJ41diCQtm6kCOMlLVmuvqqtmW44kLY9phyC+Z8XyY4DzgI8BBrAkHaVphyB+fuV6kpOAt49RkCQti6OdjvJfAMeFJekYTDsG/AdMrnqAySQ83wZcM1ZRkrQMph0D/vUVyw8Cd1TVvhHqkaSlMdUQxDApz98wmRHtZODfxyxKkpbBtE/EuAj4CPBC4CLgw0mcjlKSjsG0QxCXAt9TVfsBkmwC3g+8c6zCJGmjm/YqiOMOhu/gC0fwXknSKqY9A74+yXuBtw3rLwKuG6ckSVoOaz0T7qnA5qr6pSQ/Ajxr2PQh4C1jFydJG9laZ8D/F/hlgKp6F/AugCTfMWz7oRFrk6QNba1x3M1V9clDG4e2baNUJElLYq0APukw2x47wzokaemsFcB7k7z00MYkPw18dJySJGk5rDUG/IvAu5P8GP8RuNuBRwHPH7EuSdrwDhvAVXUv8MwkzwaeNjS/p6o+MHplkrTBTTsf8I3AjSPXIklLxbvZJKnJaAGc5PIk+5PcuqLtlCTvS/LZ4efJQ3uSvDbJbUluSXL2WHVJ0qIY8wz4CuD8Q9peBdxQVWcANwzrAM8Fzhheu4DXj1iXJC2E0QK4qm4CvnhI8w7gymH5SuDCFe1X1cTNwElJThurNklaBPMeA95cVfcMy58HNg/LW4A7V+y3b2h7mCS7kuxNsvfAgQPjVSpJI2v7Eq6qiv94ztyRvG93VW2vqu2bNm0aoTJJmo95B/C9B4cWhp8H5xi+C9i6Yr8nDm2StGHNO4D3ADuH5Z3AtSvaXzJcDXEOcP+KoQpJ2pCmnZD9iCV5G3AucGqSfcCvAq8BrklyCXAHk+fLwWRy9wuA24AHgJ8aqy5JWhSjBXBVvfgRNp23yr4FvHysWiRpEXknnCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJgawJDUxgCWpiQEsSU0MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJgawJDUxgCWpiQEsSU0MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJgawJDU5oeNDk9wO/BPwEPBgVW1PcgpwNbANuB24qKr+saM+SZqHzjPgZ1fVWVW1fVh/FXBDVZ0B3DCsS9KGtUhDEDuAK4flK4EL+0qRpPF1BXABf5zko0l2DW2bq+qeYfnzwObV3phkV5K9SfYeOHBgHrVK0ihaxoCBZ1XVXUm+CXhfkr9ZubGqKkmt9saq2g3sBti+ffuq+0jSetByBlxVdw0/9wPvBp4B3JvkNIDh5/6O2iRpXuYewEm+IcmJB5eB5wC3AnuAncNuO4Fr512bJM1TxxDEZuDdSQ5+/lur6vokfwlck+QS4A7goobaJGlu5h7AVfU54DtXaf8CcN6865GkLot0GZokLRUDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJgawJDUxgCWpiQEsSU0MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSEwNYkpoYwJLUxACWpCYGsCQ1MYAlqYkBLElNDGBJamIAS1ITA1iSmhjAktTEAJakJgawJDUxgCWpiQEsSU0MYElqYgBLUhMDWJKaGMCS1MQAlqQmBrAkNTGAJamJASxJTQxgSWpiAEtSk4UL4CTnJ/lMktuSvKq7Hkkay0IFcJLjgdcBzwXOBF6c5MzeqiRpHAsVwMAzgNuq6nNV9e/A24EdzTVJ0ihSVd01fE2SFwDnV9VPD+s/AXxvVb1ixT67gF3D6rcAnzmCjzgVuG9G5S4a+7b+bNR+gX071H1Vdf6hjSfMpp75qardwO6jeW+SvVW1fcYlLQT7tv5s1H6BfZvWog1B3AVsXbH+xKFNkjacRQvgvwTOSPKkJI8CLgb2NNckSaNYqCGIqnowySuA9wLHA5dX1adm+BFHNXSxTti39Wej9gvs21QW6ks4SVomizYEIUlLwwCWpCYbMoDXup05yaOTXD1s/3CSbQ1lHpUp+vbKJJ9OckuSG5L8p446j9S0t6An+dEklWTdXOI0Td+SXDT83j6V5K3zrvFoTfH3eHqSG5N8fPibvKCjziOV5PIk+5Pc+gjbk+S1Q79vSXL2UX1QVW2oF5Mv7/4OeDLwKOCvgDMP2efngDcMyxcDV3fXPcO+PRv4+mH5Z9dD36bp17DficBNwM3A9u66Z/g7OwP4OHDysP5N3XXPsG+7gZ8dls8Ebu+ue8q+/RfgbODWR9h+AfBHQIBzgA8fzedsxDPgaW5n3gFcOSy/EzgvSeZY49Fas29VdWNVPTCs3szkWupFN+0t6L8G/G/g3+ZZ3DGapm8vBV5XVf8IUFX751zj0ZqmbwU8flj+RuDuOdZ31KrqJuCLh9llB3BVTdwMnJTktCP9nI0YwFuAO1es7xvaVt2nqh4E7geeMJfqjs00fVvpEib/l150a/Zr+Cfe1qp6zzwLm4FpfmffDHxzkj9PcnOSh92yuqCm6dtlwI8n2QdcB/z8fEob3ZH+t7iqhboOWLOT5MeB7cB/7a7lWCU5DvhN4CebSxnLCUyGIc5l8i+Wm5J8R1X9v86iZuTFwBVV9RtJvg94c5KnVdVXuwtbBBvxDHia25m/tk+SE5j80+gLc6nu2Ex1q3aSHwQuBX64qr48p9qOxVr9OhF4GvAnSW5nMua2Z518ETfN72wfsKeqvlJVfw/8LZNAXnTT9O0S4BqAqvoQ8Bgmk9msdzOZNmEjBvA0tzPvAXYOyy8APlDDyPqCW7NvSb4LeCOT8F0vY4mH7VdV3V9Vp1bVtqraxmRs+4eram9PuUdkmr/H32dy9kuSU5kMSXxujjUerWn69g/AeQBJvo1JAB+Ya5Xj2AO8ZLga4hzg/qq654iP0v1t40jfYF7A5Czi74BLh7b/xeQ/Wpj8EbwDuA34CPDk7ppn2Lf3A/cCnxhee7prnkW/Dtn3T1gnV0FM+TsLkyGWTwOfBC7urnmGfTsT+HMmV0h8AnhOd81T9uttwD3AV5j8C+US4GeAn1nxO3vd0O9PHu3fo7ciS1KTjTgEIUnrggEsSU0MYElqYgBLUhMDWJKaGMCS1MRbkbWhJLkM+GcmE8DcVFXv761IemQGsDakqvqfYxw3yQk1mcBJOmYOQWjdS3Jpkr9N8mfAtwxtVyR5wbD8mhWT1P/6YY5zRZI3JNk7HO95Q/tPJtmT5APADUlOSfL7w/FuTvL0Yb/Lkrw5yYeSfDbJS8fvvdYzz4C1riX5biZzEJzF5O/5Y8BHV2x/AvB84FurqpKctMYhtzGZ5/YpwI1Jnjq0nw08vaq+mOS3gI9X1YVJfgC4avh8gKczmSzoG4CPJ3lPVa2LOXA1f54Ba737z8C7q+qBqvoSD58M5n4mE7i/KcmPAA8ceoBDXFNVX62qzzKZEOdbh/b3VdXBCbqfBbwZoKo+ADwhycFJx6+tqn+tqvuAG5mEubQqA1gb2jBe+wwmTz55HnD9Wm95hPV/mfYj11iXvsYA1np3E3BhkscmORH4oZUbkzwO+Maqug7478B3rnG8FyY5LslTmDzr7DOr7POnwI8Nxz8XuG84+wbYkeQxw9DHuUymbJRW5Riw1rWq+liSq5lMd7ifhwfeicC1SR7DZArBV65xyH9gMkXp45lMPfhvqzwu8DLg8iS3MBnS2Lli2y1Mhh5OBX7N8V8djtNRSoMkVwB/WFXvPMr3Xwb8c1U94pUW0koOQUhSE8+AtXSSXAq88JDmd1TVqzvq0fIygCWpiUMQktTEAJakJgawJDUxgCWpyf8H5elihEgv05sAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 360x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "\n",
    "ax = sns.displot(x='dis_prop',  data=event_dis_stat)\n",
    "#_ = ax.set_xticklabels(ax.get_xticklabels(),rotation=90)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
