{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam \n",
    "import pandas as pd "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"/Users/zluo/TestData/GENCODE/GRCh38/gencode.v31.primary_assembly.protein_coding.exon.srt.bed\",sep='\\t', names=['chrom','start','end','name','score','strand'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "fa = pysam.FastaFile(\"/Users/zluo/TestData/GENCODE/GRCh38/Sequence/GRCh38.p13.genome.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f3f03d45276d404d8db1d5dac062b188",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dask Apply:   0%|          | 0/24 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6354a0f4a07848dabafc66246807c367",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dask Apply:   0%|          | 0/24 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a69250d844ae43bea4bf7dd4d7ed1fcd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dask Apply:   0%|          | 0/24 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a1bf0db8745c4239975f0c586b663bf4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dask Apply:   0%|          | 0/24 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas\n",
    "import swifter\n",
    "import re\n",
    "def parse_gtf_attrs(feature):\n",
    "    feature_type = feature['feature']\n",
    "    attrs = feature['attribute']\n",
    "    regex = r\"(\\S+\\s\\S+;)\"\n",
    "    attrs_dict = dict()\n",
    "    for attr in re.finditer(regex,attrs):\n",
    "        attr_key = attr[1].split()[0]\n",
    "        attr_val = attr[1].split()[1].rstrip(';').strip('\"')\n",
    "        attrs_dict[attr_key] = attr_val\n",
    "\n",
    "    # parse attributes from gene features\n",
    "    # gene_id, gene_type, gene_name\n",
    "    if feature_type == 'gene':\n",
    "        feature['gene_id'] = attrs_dict['gene_id']\n",
    "        feature['gene_type'] = attrs_dict['gene_type']\n",
    "        feature['gene_name'] = attrs_dict['gene_name']\n",
    "    # parse attributes from transcript features\n",
    "    # gene_id, gene_type, gene_name\n",
    "    if feature_type == 'transcript':\n",
    "        feature['gene_id'] = attrs_dict['gene_id']\n",
    "        feature['gene_type'] = attrs_dict['gene_type']\n",
    "        feature['gene_name'] = attrs_dict['gene_name']\n",
    "        feature['transcript_id'] = attrs_dict['transcript_id']\n",
    "        if 'transcript_support_level' in attrs_dict:\n",
    "            feature['transcript_support_level'] = attrs_dict['transcript_support_level']\n",
    "        else:\n",
    "            feature['transcript_support_level'] = 'NA'\n",
    "    if feature_type == 'exon':\n",
    "        feature['gene_id'] = attrs_dict['gene_id']\n",
    "        feature['gene_type'] = attrs_dict['gene_type']\n",
    "        feature['gene_name'] = attrs_dict['gene_name']\n",
    "        feature['transcript_id'] = attrs_dict['transcript_id']\n",
    "        feature['exon_id'] = attrs_dict['exon_id']\n",
    "        feature['exon_number'] = int(attrs_dict['exon_number'])\n",
    "    if feature_type == 'CDS':\n",
    "        feature['gene_id'] = attrs_dict['gene_id']\n",
    "        feature['gene_type'] = attrs_dict['gene_type']\n",
    "        feature['gene_name'] = attrs_dict['gene_name']\n",
    "        feature['transcript_id'] = attrs_dict['transcript_id']\n",
    "        feature['exon_id'] = attrs_dict['exon_id']\n",
    "        feature['exon_number'] = int(attrs_dict['exon_number'])\n",
    "\n",
    "    return feature\n",
    "\n",
    "\n",
    "def parse_gtf(gtf_file):\n",
    "    gtf = pandas.read_csv(gtf_file, sep='\\t', comment='#', names=['seqname','source','feature','start','end','score','strand','frame','attribute'])\n",
    "    gtf_gene = gtf[gtf['feature']=='gene'].swifter.apply(parse_gtf_attrs,axis=1)\n",
    "    gtf_transcript = gtf[gtf['feature']=='transcript'].swifter.apply(parse_gtf_attrs,axis=1)\n",
    "    gtf_exon = gtf[gtf['feature']=='exon'].swifter.apply(parse_gtf_attrs,axis=1)\n",
    "    gtf_CDS = gtf[gtf['feature']=='CDS'].swifter.apply(parse_gtf_attrs,axis=1)\n",
    "    return gtf_gene, gtf_transcript, gtf_exon,gtf_CDS\n",
    "gtf_gene, gtf_ts, gtf_exon, gtf_CDS = parse_gtf(\"../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_gene.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.gene.tab',sep='\\t',index=False)\n",
    "gtf_ts.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.ts.tab',sep='\\t',index=False)\n",
    "gtf_exon.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.exon.tab',sep='\\t',index=False)\n",
    "gtf_CDS.to_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.CDS.tab',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_gene = pd.read_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.gene.tab',sep='\\t')\n",
    "gtf_ts = pd.read_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.ts.tab',sep='\\t')\n",
    "gtf_exon = pd.read_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.exon.tab',sep='\\t')\n",
    "gtf_CDS = pd.read_csv('../TestData/GENCODE/GRCh38/Annotation/gencode.v31.primary_assembly.annotation.CDS.tab',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>738</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>939275</td>\n",
       "      <td>939460</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.12\"; transcript_id \"E...</td>\n",
       "      <td>ENSG00000187634.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENST00000455979.1</td>\n",
       "      <td>ENSE00002715021.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>739</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>941144</td>\n",
       "      <td>941306</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.12\"; transcript_id \"E...</td>\n",
       "      <td>ENSG00000187634.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENST00000455979.1</td>\n",
       "      <td>ENSE00003477353.1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>740</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>942136</td>\n",
       "      <td>942251</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.12\"; transcript_id \"E...</td>\n",
       "      <td>ENSG00000187634.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENST00000455979.1</td>\n",
       "      <td>ENSE00003681266.1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>741</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>942410</td>\n",
       "      <td>942488</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.12\"; transcript_id \"E...</td>\n",
       "      <td>ENSG00000187634.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENST00000455979.1</td>\n",
       "      <td>ENSE00003675531.1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>942559</td>\n",
       "      <td>943058</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.12\"; transcript_id \"E...</td>\n",
       "      <td>ENSG00000187634.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENST00000455979.1</td>\n",
       "      <td>ENSE00002728091.1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1364091</th>\n",
       "      <td>KI270734.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>144749</td>\n",
       "      <td>144895</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277196.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000277196.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC007325.2</td>\n",
       "      <td>ENST00000621424.4</td>\n",
       "      <td>ENSE00003752738.1</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1364092</th>\n",
       "      <td>KI270734.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>143614</td>\n",
       "      <td>143789</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277196.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000277196.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC007325.2</td>\n",
       "      <td>ENST00000621424.4</td>\n",
       "      <td>ENSE00003720006.1</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1364093</th>\n",
       "      <td>KI270734.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>142194</td>\n",
       "      <td>142292</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277196.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000277196.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC007325.2</td>\n",
       "      <td>ENST00000621424.4</td>\n",
       "      <td>ENSE00003719283.1</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1364094</th>\n",
       "      <td>KI270734.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>138743</td>\n",
       "      <td>138831</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277196.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000277196.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC007325.2</td>\n",
       "      <td>ENST00000621424.4</td>\n",
       "      <td>ENSE00003751415.1</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1364095</th>\n",
       "      <td>KI270734.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>138082</td>\n",
       "      <td>138667</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277196.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000277196.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC007325.2</td>\n",
       "      <td>ENST00000621424.4</td>\n",
       "      <td>ENSE00003753010.1</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>613563 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqname   source feature   start     end score strand frame  \\\n",
       "738            chr1   HAVANA    exon  939275  939460     .      +     .   \n",
       "739            chr1   HAVANA    exon  941144  941306     .      +     .   \n",
       "740            chr1   HAVANA    exon  942136  942251     .      +     .   \n",
       "741            chr1   HAVANA    exon  942410  942488     .      +     .   \n",
       "742            chr1   HAVANA    exon  942559  943058     .      +     .   \n",
       "...             ...      ...     ...     ...     ...   ...    ...   ...   \n",
       "1364091  KI270734.1  ENSEMBL    exon  144749  144895     .      -     .   \n",
       "1364092  KI270734.1  ENSEMBL    exon  143614  143789     .      -     .   \n",
       "1364093  KI270734.1  ENSEMBL    exon  142194  142292     .      -     .   \n",
       "1364094  KI270734.1  ENSEMBL    exon  138743  138831     .      -     .   \n",
       "1364095  KI270734.1  ENSEMBL    exon  138082  138667     .      -     .   \n",
       "\n",
       "                                                 attribute  \\\n",
       "738      gene_id \"ENSG00000187634.12\"; transcript_id \"E...   \n",
       "739      gene_id \"ENSG00000187634.12\"; transcript_id \"E...   \n",
       "740      gene_id \"ENSG00000187634.12\"; transcript_id \"E...   \n",
       "741      gene_id \"ENSG00000187634.12\"; transcript_id \"E...   \n",
       "742      gene_id \"ENSG00000187634.12\"; transcript_id \"E...   \n",
       "...                                                    ...   \n",
       "1364091  gene_id \"ENSG00000277196.4\"; transcript_id \"EN...   \n",
       "1364092  gene_id \"ENSG00000277196.4\"; transcript_id \"EN...   \n",
       "1364093  gene_id \"ENSG00000277196.4\"; transcript_id \"EN...   \n",
       "1364094  gene_id \"ENSG00000277196.4\"; transcript_id \"EN...   \n",
       "1364095  gene_id \"ENSG00000277196.4\"; transcript_id \"EN...   \n",
       "\n",
       "                    gene_id       gene_type   gene_name      transcript_id  \\\n",
       "738      ENSG00000187634.12  protein_coding      SAMD11  ENST00000455979.1   \n",
       "739      ENSG00000187634.12  protein_coding      SAMD11  ENST00000455979.1   \n",
       "740      ENSG00000187634.12  protein_coding      SAMD11  ENST00000455979.1   \n",
       "741      ENSG00000187634.12  protein_coding      SAMD11  ENST00000455979.1   \n",
       "742      ENSG00000187634.12  protein_coding      SAMD11  ENST00000455979.1   \n",
       "...                     ...             ...         ...                ...   \n",
       "1364091   ENSG00000277196.4  protein_coding  AC007325.2  ENST00000621424.4   \n",
       "1364092   ENSG00000277196.4  protein_coding  AC007325.2  ENST00000621424.4   \n",
       "1364093   ENSG00000277196.4  protein_coding  AC007325.2  ENST00000621424.4   \n",
       "1364094   ENSG00000277196.4  protein_coding  AC007325.2  ENST00000621424.4   \n",
       "1364095   ENSG00000277196.4  protein_coding  AC007325.2  ENST00000621424.4   \n",
       "\n",
       "                   exon_id  exon_number  \n",
       "738      ENSE00002715021.1            1  \n",
       "739      ENSE00003477353.1            2  \n",
       "740      ENSE00003681266.1            3  \n",
       "741      ENSE00003675531.1            4  \n",
       "742      ENSE00002728091.1            5  \n",
       "...                    ...          ...  \n",
       "1364091  ENSE00003752738.1           11  \n",
       "1364092  ENSE00003720006.1           12  \n",
       "1364093  ENSE00003719283.1           13  \n",
       "1364094  ENSE00003751415.1           14  \n",
       "1364095  ENSE00003753010.1           15  \n",
       "\n",
       "[613563 rows x 15 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_gene_coding = gtf_gene[gtf_gene['gene_type'] == 'protein_coding']\n",
    "gtf_ts_coding = gtf_ts[gtf_ts['gene_id'].isin(gtf_gene_coding['gene_id'])]\n",
    "gtf_ts_coding_tsl1and2 = gtf_ts_coding[gtf_ts_coding['transcript_support_level'].isin(['1','2',1,2])]\n",
    "gtf_exon_coding_tsl1and2 = gtf_exon[gtf_exon['transcript_id'].isin(gtf_ts_coding_tsl1and2['transcript_id'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6\n"
     ]
    }
   ],
   "source": [
    "gtf_gene_coding\n",
    "gtf_ts_coding_tsl1and2_sample = gtf_ts_coding_tsl1and2.sample(n=1000, random_state=1992)\n",
    "for idx, row in gtf_ts_coding_tsl1and2_sample.iterrows():\n",
    "    ts_id = row['transcript_id']\n",
    "    exons  = gtf_exon_coding_tsl1and2[gtf_exon_coding_tsl1and2['transcript_id'] == ts_id]\n",
    "    exons = exons.sort_values('exon_number')\n",
    "    exon_cnt = exons['exon_number'].max()\n",
    "    for idx_exon, exon in exons.iterrows():\n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
